Update language database

utelle opened the pull request:
- List of languages updated to include most locales supported by Windows 10
- Added attributes to each language entry: BCP 47-like locale tag, reference to canonical name of default region for generic language entries, description in native language
- Added data file with list of language script identifiers and aliases based on ISO 15924
- Updated Python script genlang.py to handle the new attributes and data
- Implemented most static methods of wxLocale in wxUILocale and redirected wxLocale methods to the new wxUILocale methods
- Updated German translation of internat sample

utelle committed:
Update language database

- List of languages updated to include most locales supported by Windows 10
- Added attributes to each language entry: BCP 47-like locale tag, reference to canonical name of default region for generic language entries, description in native language
- Added data file with list of language script identifiers and aliases based on ISO 15924
- Updated Python script genlang.py to handle the new attributes and data
- Implemented most static methods of wxLocale in wxUILocale and redirected wxLocale methods to the new wxUILocale methods
- Updated German translation of internat sample

utelle commented:
The intended goal is to support more locales in conjunction with the new wxUILocale class. The wxUILocale class will be extended to retrieve some additional information items like localized names (locale, language, country, ...). For Windows versions greater or equal Windows 7 locale identification and selection will be based on BCP 47-like language tags instead of the deprecated numeric LCID language identifiers (which support only a limited number of locales).

This PR is a first step to achieve the above goal. It updates the language database and implements the language database related static methods of wxLocale in wxUILocale. The implementation bodies were not changed. The internat sample works as before.

vadz commented on misc/languages/langtabl.txt:
```diff
@@ -1,236 +1,879 @@
-wxLANGUAGE_YORUBA                      yo     -                -                                   LTR    "Yoruba"
-wxLANGUAGE_ZHUANG                      za     -                -                                   LTR    "Zhuang"
-wxLANGUAGE_ZULU                        zu     -                -                                   LTR    "Zulu"
+wxLANGUAGE_ABKHAZIAN                                    ab             ab                        -              -    -    LTR "Abkhazian","\320\260\322\247\321\201\321\203\320\260 \320\261\321\213\320\267\321\210\323\231\320\260"
```
Are octal escapes from the original data source? If not, I think using UTF-8 for this file could be better.

utelle replied:
No, the octal escapes are not from the original data. The original data are in UTF-8. However, since the data are used to generate a C++ source code file, I prefer hexadecimal or octal escapes to keep the character set of the source file to pure ASCII. For example, Visual Studio sometimes does weird things, if the source file encoding is not ASCII. Here I used octal escapes, because they are as compact as hex escapes and don't introduce problems when the following character in the string happens to be a valid hex digit.

vadz commented on include/wx/uilocale.h:
```diff
@@ -138,11 +138,68 @@ class WXDLLIMPEXP_BASE wxUILocale
     explicit wxUILocale(wxUILocaleImpl* impl = NULL) : m_impl(impl) { }
 
+    // copy default table of languages from global static array to
+    // m_languagesInfo, called by InitLanguagesDB
```
I think the comment needs to be updated as it says that this function is called by itself.

utelle replied:
Actually, I simply copied the existing comment without further thoughts. In fact, the comment was wrong even for the previous implementation. I will try to find a better wording.

vadz commented on include/wx/uilocale.h:
```diff
@@ -138,11 +138,68 @@ class WXDLLIMPEXP_BASE wxUILocale
+
+    // get the encoding used by default for text on this system, returns
+    // wxFONTENCODING_SYSTEM if it couldn't be determined
+    static wxFontEncoding GetSystemEncoding();
```
I don't think we need this function any more, font encodings are only useful in non-Unicode builds and we'll be dropping support for them soon, so why bother adding new functions used by them only.

utelle replied:
Ok, I wasn't aware of that. So, I will revert the change.

vadz commented on include/wx/uilocale.h:
```diff
@@ -138,11 +138,68 @@ class WXDLLIMPEXP_BASE wxUILocale
+
+    // get the string describing the system encoding, return empty string if
+    // couldn't be determined
+    static wxString GetSystemEncodingName();
```
We probably don't need this one neither.

utelle replied:
Ok, I will revert this change, too.

vadz commented on interface/wx/intl.h:
```diff
@@ -33,9 +33,17 @@ struct wxLanguageInfo
     /// language info structure.
     int Language;
 
+    /// Tag for locale in BCP 47-like notation.
```
It should be documented that this, and the other new, field(s) is/are only available since wx 3.1.6.

utelle replied:
Ok, I will adjust the comments.

vadz commented on src/common/intl.cpp:
```diff
@@ -223,264 +225,22 @@ const wxLanguageInfos& wxGetLanguageInfos()
 }
 
-
-void wxLocale::DoCommonInit()
```
I think this code was just moved, without changes, but GitHub UI doesn't have the equivalent of Git `--color-moved`, so I'm not sure -- please let me know if there were any changes worth being reviewed here.

utelle replied:
Yes, the code was simply _moved_ **without any changes**. 

Maybe a later PR will introduce changes, but no changes for now.

vadz reviewed (commented):
Thanks for your changes!

I'm not sure if the changes to `wx[UI]Locale` were supposed to be in this PR, as I thought it was meant to include only the language DB changes, but they all look good to me anyhow.

The only other thing I see is that `misc/languages/README` should be updated to mention the new `scripttabl.txt` file and, ideally, also explain, even if very briefly, how should this file and `langtabl.txt` be updated in the future.

TIA!

vadz commented:
It certainly looks that locale loading got broken under Mac by these changes, but I don't see why immediately and don't have time to debug this right now, so this will have to wait a bit until I (or somebody else) do(es).

utelle commented on include/wx/localedefs.h:
```diff
@@ -82,12 +82,15 @@ enum wxLocaleInfo
     int Language;                   // wxLanguage id
+    wxString LocaleTag;             // Tag of locale in BCP 47-like notation
     wxString CanonicalName;         // Canonical name, e.g. fr_FR
+    wxString CanonicalRef;          // Canonical name reference, e.g. fr_FR for fr
```
In principle, I'm fine with your suggested wording.

However, I added `CanonicalRef` only for those entries of the previous language database, for which no entry with just the language code existed. I did that, so that the translation loading method can be adjusted to look not only in the directory specified by the language code alone (for example **fr**), but also in the directory specified by language and region (for example **fr_FR**).

I'm not sure whether we really need to do that, but I made these provisions to be on the safe side. BTW, the change to the translation loading code has still to be done.

utelle commented:
> It certainly looks that locale loading got broken under Mac by these changes, but I don't see why immediately and don't have time to debug this right now, so this will have to wait a bit until I (or somebody else) do(es).

AFAICT there were no code changes that could cause the locale loading to fail. I only copied the implementation of several static wxLocale methods to static wxUILocale methods.

The only change that possibly could matter is in the language data. The canonical name for the French locale is now **fr** (instead of **fr_FR** as in the previous version) and the code was not yet changed to make use of the attribute `CanonicalRef`. Could this cause the problem??? It shouldn't as not all locale names have a regional part, and it works on the other platforms, but macOS might be different.

I will try to debug this myself, however, I already struggled a couple of days to get `gdb` to work in my macOS VM ... without success up to now.

vadz commented:
> The only change that possibly could matter is in the language data. The canonical name for the French locale is now **fr** (instead of **fr_FR** as in the previous version) and the code was not yet changed to make use of the attribute `CanonicalRef`. Could this cause the problem???

I don't see how could this happen... Even if the test tried loading `fr_FR` locale instead of `fr` (which it has), it should end up with `fr` one due the fallback logic. Sorry, this really needs to be debugged :-(

> I will try to debug this myself, however, I already struggled a couple of days to get `gdb` to work in my macOS VM ... without success up to now.

I advise using lldb under macOS instead. [This page](https://lldb.llvm.org/use/map.html) could be helpful -- at least I used to keep it open all the time when starting to work with lldb.

> > The only other thing I see is that `misc/languages/README` should be updated to mention the new `scripttabl.txt` file and, ideally, also explain, even if very briefly, how should this file and `langtabl.txt` be updated in the future.
> 
> Sorry, just an oversight. I will add a paragraph to the readme file.

Great, TIA!

utelle commented:
> > The only change that possibly could matter is in the language data. The canonical name for the French locale is now **fr** (instead of **fr_FR** as in the previous version) and the code was not yet changed to make use of the attribute `CanonicalRef`. Could this cause the problem???
> 
> I don't see how could this happen... Even if the test tried loading `fr_FR` locale instead of `fr` (which it has), it should end up with `fr` one due the fallback logic. Sorry, this really needs to be debugged :-(

I managed to perform a first debug session using the Xcode project file of the minimal sample. I added the relevant code from the tests (essentially copying the code from `IntlTestCase::setUp()` into `minimal.cpp`) and stepped through it in debug mode ... maybe not the most elegant way to do this, but it worked. 😄 

The test code calls method `wxLocale::Init`:
```C++
bool loaded = m_locale->Init(wxLANGUAGE_FRENCH, wxLOCALE_DONT_LOAD_DEFAULT);
CPPUNIT_ASSERT( loaded );
```
This call returns `false` - resulting in the failure of the assertion.

Method `Init`calls internally `wxUILocale::UseLanguage` ... and this works and returns _true_. However, the next step is to call `info->TrySetLocale()` ... and this fails - `TrySetLocale` internally calls `wxSetLocale` which in turn calls `setlocale`. And here is the culprit: `setlocale(LC_ALL, "fr_FR")` succeeds, while `setlocale(LC_ALL, "fr")` fails. 💥 

Interestingly, `setlocale(LC_ALL, "fr")` seems to work without issues on Linux/Unix platforms. So the big question is, must `setlocale` on macOS be called always with an identifier which includes the region code?

The previous language database contains quite a few identifiers without region code. I checked my macOS VM and `locale -a` lists at least 3 locales for which the wx language database identifiers don't include a region code: Amharic (**am** vs **am_ET**), Armenian (**hy** vs **hy_AM**), and Kazakh (**kk** vs **kk_KZ**). I' ll try to test tomorrow, what happens if I choose one of those languages.

utelle commented:
> The previous language database contains quite a few identifiers without region code. I checked my macOS VM and `locale -a` lists at least 3 locales for which the wx language database identifiers don't include a region code: Amharic (**am** vs **am_ET**), Armenian (**hy** vs **hy_AM**), and Kazakh (**kk** vs **kk_KZ**). I' ll try to test tomorrow, what happens if I choose one of those languages.

Following the results of my tests:

The old language database contains entries for **"am"**, **"hy"**, and **"kk"**.
The new language database contains entries for **"am"**, **"hy"**, **"kk"**, and **"am_ET"**.

Calling `setlocale` with locale string
- **"am"** _fails_
- **"hy"** _fails_
- **"kk"** _fails_
- **"am_ET"** _succeeds_

So, it seems `setlocale` on macOS **requires** the presence of a _region_ identifier in the locale string.

It should be mentioned that the internal check for available locales in the `Create` method in `uilocale.mm` succeeds for **all** of the above listed locales. I get the impression that the `setlocale` implementation on macOS is somehow broken. The question is, what we should do about that.

vadz commented:
Sorry for missing this previously, I do know that `setlocale("fr")` doesn't work under Mac (if you check `locale -a | grep '^fr'` output you can see that it's indeed not in the list of supported locales), but I still don't understand what has changed here. Did we use `fr_FR` before? Or did we just not call `setlocale()` under Mac at all?

To be clear, Unix locale support should _not_ be used under Mac, it's present, but known to be subtly broken (e.g. uses wrong decimal separator for `de_CH` and changing it to any locale using comma as decimal separator breaks menu appearance for the GUI applications due to a bug in Apple CoreSVG framework), which is why `wxUILocale` only uses Core Foundation functions and using `wxLocale` there is not recommended. But the test should still continue to work, if it did before...

utelle committed:
Update language database (some adjustments)

- Adjust comments for new attributes in wxLanguageInfo and a few other comments
- Adjust interface descriptions
- Revert changes regarding methods wxLocale::GetSystemEncoding and wxLocale::GetSystemEncodingName
- Adjusted method wxLanguageInfo::TrySetLocale to use CanonicalRef if given for a CanonicalName without region identifier
- Update README in misc/languages

utelle commented:
> Sorry for missing this previously, I do know that `setlocale("fr")` doesn't work under Mac

The point is: it **should** work. **"fr"** is obviously found in the list of supported locales (see method `wxUILocaleImplCF::Create` in `uilocale.mm`). If the parameter `locale` does not contain a region identifier and/or a script identifier, the function `setlocale` should nevertheless select the locale based on the default region (and default script) for the language. This works for other Linux/Unix platforms, but not for macOS.

> (if you check `locale -a | grep '^fr'` output you can see that it's indeed not in the list of supported locales),

This is true, but it is also true for my OpenSuSE Linux installation. Nevertheless, `setlocale` works as expected and selects **"fr_FR"** for the locale parameter **"fr"**.

> but I still don't understand what has changed here. Did we use `fr_FR` before? Or did we just not call `setlocale()` under Mac at all?

The previous language database did not contain an entry for **"fr"**, but an entry for **"fr_FR"**. The search function - when called with **"fr"** - checks the canonical name whether it starts with **"fr"** and selects the first entry for which the prefix matches. Therefore **"fr_FR"** in the call to `setlocale`.

The new language database contains separate entries for **fr"** and **"fr_FR"**, because locale identifiers **"fr"** and **"fr_FR"** are both valid identifiers. **"fr"** denotes a generic identifier, not restricted to region **"FR"** alone. You have a similar situation for Arabic, where **"ar"** denotes a generic identifier for "Arabic (World)". 

The previous language database contains several other languages for which the language identifier does not include a region (like the ones mentioned in my previous post: **"am"**, **"hy"**, **"kk"**) and I fear that `setlocale`on macOS doesn't work for any of them (as tested for those that were identified as supported in my macOS VM).

I adjusted method `wxLanguageInfo::TrySetLocale`, so that the attribute `CanonicalRef` is used when available. I committed the changes, and now the tests for all mac builds succeed.

Maybe we need to adjust the entries for languages without region identifier in `langtabl.txt`, where possible, by adding an appropriate value for `CanonicalRef`. This might not be possible for all such entries, though.

> To be clear, Unix locale support should _not_ be used under Mac, it's present, but known to be subtly broken (e.g. uses wrong decimal separator for `de_CH` and changing it to any locale using comma as decimal separator breaks menu appearance for the GUI applications due to a bug in Apple CoreSVG framework), which is why `wxUILocale` only uses Core Foundation functions and using `wxLocale` there is not recommended. But the test should still continue to work, if it did before...

... and it does after the recent changes.

However, what should be done about calls to `setlocale`? For example, for testing (on Windows) I used `wxLanguageInfo::GetLocaleName` in the internat sample and got an assertion from `wxLocale::GetInfo` about a _mismatch between C/C++ and Windows locale_. Shouldn't be calls to `setlocale` via `wxSetLocale` avoided, not only on macOS, but also 
on other platforms?

Please review my latest changes whenever you find time to do it. TIA!

vadz commented on interface/wx/uilocale.h:
```diff
@@ -189,6 +189,71 @@ class wxUILocale
+
+        See GetLanguageInfo() for a remark about special meaning of @c wxLANGUAGE_DEFAULT.
+
+        @since 2.9.1
```
This is trivial, but I think we should remove it, as this entire class is only available since 3.1.6.
```suggestion
        @since 2.9.1
```

utelle replied:
Ah, sorry. Just a careless copy&paste issue. You are absolutely right that this line should be removed.

vadz commented on interface/wx/intl.h:
```diff
@@ -33,9 +33,25 @@ struct wxLanguageInfo
     wxString CanonicalName;
 
+    /**
+        Canonical reference including region, if the name specifies
```
Really a nitpick, but the brief description ought to end with a period for Doxygen, i.e.
```suggestion
        Canonical reference including region.
        
        If the name specifies
```

utelle replied:
Thanks. I'll try to keep this in mind for the next modifications.

vadz commented on interface/wx/intl.h:
```diff
@@ -45,9 +61,16 @@ struct wxLanguageInfo
     wxString Description;
 
+    /**
+        Human-readable name of the language in native language.
```
Another nitpick, but "native" might be a bit unclear here because it could be interpreted as meaning the current system language, so I'd change it to
```suggestion
        Human-readable name of the language in this language itself.
```

utelle replied:
I'm fine with the new wording.

vadz commented on src/common/intl.cpp:
```diff
@@ -167,7 +159,7 @@ const char* wxLanguageInfo::TrySetLocale() const
 const char* wxLanguageInfo::TrySetLocale() const
 {
-    return wxSetlocale(LC_ALL, CanonicalName);
+    return wxSetlocale(LC_ALL, (CanonicalRef.empty()) ? CanonicalName : CanonicalRef);
```
Very minor, but why extra parenthesis?
```suggestion
    return wxSetlocale(LC_ALL, CanonicalRef.empty() ? CanonicalName : CanonicalRef);
```

utelle replied:
> Very minor, but why extra parenthesis?

Well, just a habit to include the condition of the conditional operator in parentheses. For a single term they are superfluous, of course.

vadz reviewed (commented):
Thanks for the explanations, I understand what happened now and think that the fix is fine and am ready to merge it -- I can apply the minor fixes proposed in the latest version of the review myself if you don't mind, just let me know.

But now I don't understand the `setlocale()` problem. If it's related to this PR (i.e. if it affects/worsens it in any way), could you please explain what exactly it is, i.e. what should be done to trigger the assert? If it's not directly related to it, we could merge this PR first and discuss that other problem on wx-dev later.

Thanks again!

utelle commented:
> Thanks for the explanations, I understand what happened now and think that the fix is fine and am ready to merge it -- I can apply the minor fixes proposed in the latest version of the review myself if you don't mind, just let me know.

Please allow a couple of days, before merging the PR. I'd like to adjust the attribute `CanonicalRef` for some language entries, so that selecting them will work on macOS (if they are supported). This change may not be necessary in the long run, but it does no harm either. In that course I can also apply your proposed modifications.

> But now I don't understand the `setlocale()` problem. If it's related to this PR (i.e. if it affects/worsens it in any way),

I don't think that it is related to this PR in any way, but exists at least since `wxUILocale` was introduced. It may even exist in prior wx versions, but I haven't tested this yet.

> could you please explain what exactly it is, i.e. what should be done to trigger the assert?

Triggering the assert is simple. Just add the line

```C++
    wxString localeName = info->GetLocaleName();
```
in event handler `MyFrame::OnTestLocaleAvail`of the _internat_ sample, after the `wxLanguageInfo` instance `info` has been created successfully. Then start the sample and select the test for availability of a locale. When asked for a locale identifier enter "fr_FR" or "de_DE" or whatever. Before the result of the check is displayed the assert debug message will be displayed.

> If it's not directly related to it, we could merge this PR first and discuss that other problem on wx-dev later.

The problem exists definitely already for wx master without the PR.

utelle committed:
Update language database (adjust CanonicalRef)

- adjust attribute CanonicalRef for several entries where CanonicalName has no region identifier
- fix minor typos

utelle committed:
Remove trailing whitespace

utelle commented:
> Please allow a couple of days, before merging the PR. I'd like to adjust the attribute `CanonicalRef` for some language entries, so that selecting them will work on macOS (if they are supported).

In the meantime I applied the intended adjustments, and (hopefully) fixed all typos. So, from my side the PR would be ready for merging.

vadz commented:
Thanks for the updates! I still didn't have time to look at the other problem, but as you say that it is unrelated, I'm going to (squash) merge this already to make it more convenient for you to continue working on this and will hopefully get to that problem later.

Thanks again!

utelle commented:
Thanks for merging the PR.

Regarding the problem with the assertions, I think it has to do with the fact that method `wxLanguageInfo::GetLocaleName` calls internally `wxLanguageInfo::TrySetLocale()` which in turn calls `wxSetlocale`.  This somehow interferes with the expectation that still the C locale is active. I haven't found out yet what exactly is going on, but actually `wxLanguageInfo::GetLocaleName` leaves behind a global locale that is **not** the C locale - at least on my development system.

IMHO the method `wxLanguageInfo::GetLocaleName` should use the new `wxUILocale` class to determine the locale name. That would avoid interference with the global locale.
